ui/admin/app/components/chat/ChatActions.tsx

```diff
@@ -0,0 +1,192 @@
+import { LibraryIcon, WrenchIcon } from "lucide-react";
+import { useMemo } from "react";
+
+import { Agent } from "~/lib/model/agents";
+import { KnowledgeFile } from "~/lib/model/knowledge";
+import { cn } from "~/lib/utils";
+
+import { TypographyMuted, TypographySmall } from "~/components/Typography";
+import { ToolEntry } from "~/components/agent/ToolEntry";
+import { useChat } from "~/components/chat/ChatContext";
+import {
+    useOptimisticThread,
+    useThreadAgents as useThreadAgent,
+    useThreadKnowledge,
+} from "~/components/chat/thread-helpers";
+import { Button } from "~/components/ui/button";
+import {
+    Popover,
+    PopoverContent,
+    PopoverTrigger,
+} from "~/components/ui/popover";
+import { Switch } from "~/components/ui/switch";
+
+export function ChatActions({ className }: { className?: string }) {
+    const { threadId } = useChat();
+
+    const { data: knowledge } = useThreadKnowledge(threadId);
+    const { data: agent } = useThreadAgent(threadId);
+    const { thread, updateThread } = useOptimisticThread(threadId);
+
+    const tools = thread?.tools;
+
+    return (
+        <div className={cn("w-full flex items-center", className)}>
+            <div className="flex items-center gap-2">
+                <ToolsInfo
+                    tools={tools ?? []}
+                    onChange={(tools) => updateThread({ tools })}
+                    agent={agent}
+                    disabled={!thread}
+                />
+
+                <KnowledgeInfo knowledge={knowledge ?? []} disabled={!thread} />
+            </div>
+        </div>
+    );
+}
+
+type ToolItem = {
+    tool: string;
+    isToggleable: boolean;
+    isEnabled: boolean;
+};
+
+function ToolsInfo({
+    tools,
+    className,
+    agent,
+    disabled,
+    onChange,
+}: {
+    tools: string[];
+    className?: string;
+    agent: Nullish<Agent>;
+    disabled?: boolean;
+    onChange: (tools: string[]) => void;
+}) {
+    const toolItems = useMemo<ToolItem[]>(() => {
+        if (!agent)
+            return tools.map((tool) => ({
+                tool,
+                isToggleable: false,
+                isEnabled: true,
+            }));
+
+        const agentTools = (agent.tools ?? []).map((tool) => ({
+            tool,
+            isToggleable: false,
+            isEnabled: true,
+        }));
+
+        const { defaultThreadTools, availableThreadTools } = agent ?? {};
+
+        const toggleableTools = [
+            ...(defaultThreadTools ?? []),
+            ...(availableThreadTools ?? []),
+        ].map((tool) => ({
+            tool,
+            isToggleable: true,
+            isEnabled: tools.includes(tool),
+        }));
+
+        return [...agentTools, ...toggleableTools];
+    }, [tools, agent]);
+
+    const handleToggleTool = (tool: string, checked: boolean) => {
+        console.log("toggle tool", tool, checked);
```

Comment:
nit: alright to remove this console.log?

Reply:
done